browser/js/dashboard/gridbox/WidgetSettingsFactory.js

```diff
@@ -0,0 +1,36 @@
+app.factory('WidgetSettingsFactory', function($interval, DashboardFactory){
+  var factory = {};
+
+  factory.newSetKeys = function(dataSource){
+    return DashboardFactory.getDataSource(dataSource)
+      .then(function(data){
+        var realData = factory.findDataToGraph(data);
+        if (!realData) {
+          realData = [data];
+        }
+        var dataInNVD3Format = [{
+          values:realData,
+          key: "this works",
+          color: '#ff7f0e'
+        }];
+        return [dataInNVD3Format,Object.keys(realData[0])]
```

Comment:
Maybe you could return an object instead? Something like `{format: x, keys: y}` instead of `[x, y]`. This would be so that the data you're working with has a more descriptive format.